## Part 1: Preprocessing

In [128]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [129]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [130]:
attrition_df.dtypes


Age                          int64
Attrition                   object
BusinessTravel              object
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EnvironmentSatisfaction      int64
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
NumCompaniesWorked           int64
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StockOptionLevel             int64
TotalWorkingYears            int64
TrainingTimesLastYear        int64
WorkLifeBalance              int64
YearsAtCompany               int64
YearsInCurrentRole           int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

In [131]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition','Department']]
y_df.head()
y_df.dtypes

Attrition     object
Department    object
dtype: object

In [132]:
# Create a list of at least 10 column names to use as X data
X_data =[
'Education',
'Age',
'DistanceFromHome',
'JobSatisfaction', 
'OverTime',
'StockOptionLevel',
'WorkLifeBalance',
'YearsAtCompany',
'YearsSinceLastPromotion',
'NumCompaniesWorked']


# Create X_df using your selected columns
X_df=attrition_df[X_data]

# Show the data types for X_df

print(X_df.dtypes)

Education                   int64
Age                         int64
DistanceFromHome            int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
NumCompaniesWorked          int64
dtype: object


In [133]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, test_size=0.2, random_state=42)


In [134]:
X_df['OverTime']

0       Yes
1        No
2       Yes
3       Yes
4        No
       ... 
1465     No
1466     No
1467    Yes
1468     No
1469     No
Name: OverTime, Length: 1470, dtype: object

In [135]:
# Convert your X data to numeric data types however you see fit
X_df = X_df.astype({
    'Education': 'int32',
    'Age': 'int32',
    'DistanceFromHome': 'float32',
    'JobSatisfaction': 'int32',
    'OverTime': 'category',  
    'StockOptionLevel': 'int32',
    'WorkLifeBalance': 'int32',
    'YearsAtCompany': 'int32',
    'YearsSinceLastPromotion': 'int32',
    'NumCompaniesWorked': 'int32'
})

# Show the new data types
print("New data types after conversion:")
print(X_df.dtypes)
print("\n")

# Check for any remaining non-numeric columns
non_numeric = X_df.select_dtypes(exclude=[np.number]).columns
if len(non_numeric) > 0:
    print(f"Non-numeric columns: {list(non_numeric)}")
else:
    print("All columns are numeric except for 'OverTime' which is categorical.")

# Display the first few rows of the processed DataFrame
print("\nFirst few rows of processed X_df:")
print(X_df.head())

New data types after conversion:
Education                     int32
Age                           int32
DistanceFromHome            float32
JobSatisfaction               int32
OverTime                   category
StockOptionLevel              int32
WorkLifeBalance               int32
YearsAtCompany                int32
YearsSinceLastPromotion       int32
NumCompaniesWorked            int32
dtype: object


Non-numeric columns: ['OverTime']

First few rows of processed X_df:
   Education  Age  DistanceFromHome  JobSatisfaction OverTime  \
0          2   41               1.0                4      Yes   
1          1   49               8.0                2       No   
2          2   37               2.0                3      Yes   
3          4   33               3.0                3      Yes   
4          1   27               2.0                2       No   

   StockOptionLevel  WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0                 0                1              

In [136]:
# Encode 'OverTime' 


if 'OverTime' in X_train.columns:
    X_train['OverTime'] = X_train['OverTime'].map({'No': 0, 'Yes': 1})
    X_test['OverTime'] = X_test['OverTime'].map({'No': 0, 'Yes': 1})

In [137]:
X_df['OverTime'] = X_df['OverTime'].map({'No': 0, 'Yes': 1, np.nan: -1})
X_df.dtypes

Education                     int32
Age                           int32
DistanceFromHome            float32
JobSatisfaction               int32
OverTime                   category
StockOptionLevel              int32
WorkLifeBalance               int32
YearsAtCompany                int32
YearsSinceLastPromotion       int32
NumCompaniesWorked            int32
dtype: object

In [138]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [139]:
from sklearn.preprocessing import OneHotEncoder
# Create a OneHotEncoder for the Department column

dept_encoder = OneHotEncoder()
# Fit the encoder to the training data
dept_encoder.fit(y_df[['Department']])

# Create two new variables by applying the encoder
# to the training and testing data
y_train_dept_encoded = dept_encoder.transform(y_train[['Department']])
y_test_dept_encoded = dept_encoder.transform(y_test[['Department']])


In [140]:
# Create a OneHotEncoder for the Attrition column
attrition_enconder = OneHotEncoder()
# Fit the encoder to the training data
attrition_enconder.fit(y_df[['Attrition']])

# Create two new variables by applying the encoder
# to the training and testing data
y_train_att_encoded = attrition_enconder.transform(y_train[['Attrition']])
y_test_att_encoded = attrition_enconder.transform(y_test[['Attrition']])


## Create, Compile, and Train the Model

In [141]:
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Dense, BatchNormalization, Dropout
# Find the number of columns in the X training data
num_features = X_train.shape[1]
print(f"Number of input features: {num_features}")
# Create the input layer
inputs = Input(shape=(num_features,), name='input_layer')

# Create at least two shared layers
x = Dense(64, activation='relu', name='dense_1')(inputs)
x = Dense(32, activation='relu', name='dense_2')(x)

Number of input features: 10


In [142]:
temp_model = Model(inputs=inputs, outputs=x)
temp_model.summary()

Model: "functional_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 10)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 64)             │           704 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 32)             │         2,080 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,784 (10.88 KB)

 Trainable params: 2,784 (10.88 KB)

 Non-trainable params: 0 (0.00 B)

In [143]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
dept_hidden = Dense(16, activation='relu', name='dept_hidden')(x)
num_departments = len(y_train['Department'].unique())


# Create the output layer
dept_output = Dense(num_departments, activation='softmax', name='dept_output')(dept_hidden)


In [144]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = Dense(8, activation='relu', name='attrition_hidden')(x)


# Create the output layer
attrition_output = Dense(1, activation='sigmoid', name='attrition_output')(attrition_hidden)



In [147]:
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
# Create the model
model = Model(inputs=inputs, outputs=[dept_output, attrition_output], name='employee_prediction_model')


# Compile the model
model.compile(optimizer=Adam(learning_rate=0.001),
              loss={'dept_output': 'categorical_crossentropy',
                    'attrition_output': 'binary_crossentropy'},
              loss_weights={'dept_output': 1.0, 'attrition_output': 1.0},
              metrics={'dept_output': 'accuracy',
                       'attrition_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "employee_prediction_model"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 10)        │          0 │ -                 │
│ (InputLayer)        │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_1 (Dense)     │ (None, 64)        │        704 │ input_layer[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_2 (Dense)     │ (None, 32)        │      2,080 │ dense_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dept_hidden (Dense) │ (None, 16)        │        528 │ dense_2[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attrition_hidden    │ (None, 8)         │        264 │ dense_2[0][0]     │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dept_output (Dense) │ (None, 3)         │         51 │ dept_hidden[0][0] │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ attrition_output    │ (None, 1)         │          9 │ attrition_hidden… │
│ (Dense)             │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 3,636 (14.20 KB)

 Trainable params: 3,636 (14.20 KB)

 Non-trainable params: 0 (0.00 B)

In [150]:
# Train the model
y_train_dept = pd.get_dummies(y_train['Department'])  # One-hot encode Department
y_train_attrition = y_train['Attrition'].map({'No': 0, 'Yes': 1})  # Binary encode Attrition

y_test_dept = pd.get_dummies(y_test['Department'])  # One-hot encode Department
y_test_attrition = y_test['Attrition'].map({'No': 0, 'Yes': 1})  # Binary encode Attrition
history = model.fit(
    X_train_scaled, 
    [y_train_dept, y_train_attrition],
    validation_split=0.2,
    epochs=100,
    batch_size=32,
    verbose=1
)

Epoch 1/100


C:\Users\austi\AppData\Roaming\Python\Python310\site-packages\keras\src\losses\losses.py:27: SyntaxWarning: In loss categorical_crossentropy, expected y_pred.shape to be (batch_size, num_classes) with num_classes > 1. Received: y_pred.shape=(None, 1). Consider using 'binary_crossentropy' if you only have 2 classes.
  return self.fn(y_true, y_pred, **self._fn_kwargs)


30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 9ms/step - attrition_output_accuracy: 0.8547 - attrition_output_loss: 0.0000e+00 - dept_output_accuracy: 0.5702 - dept_output_loss: 0.6439 - loss: 0.6439 - val_attrition_output_accuracy: 0.7966 - val_attrition_output_loss: 0.0000e+00 - val_dept_output_accuracy: 0.6314 - val_dept_output_loss: 0.5513 - val_loss: 0.5545
Epoch 2/100
 1/30 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - attrition_output_accuracy: 0.8750 - attrition_output_loss: 0.0000e+00 - dept_output_accuracy: 0.7500 - dept_output_loss: 0.5190 - loss: 0.5190

C:\Users\austi\AppData\Roaming\Python\Python310\site-packages\keras\src\losses\losses.py:27: SyntaxWarning: In loss categorical_crossentropy, expected y_pred.shape to be (batch_size, num_classes) with num_classes > 1. Received: y_pred.shape=(None, 1). Consider using 'binary_crossentropy' if you only have 2 classes.
  return self.fn(y_true, y_pred, **self._fn_kwargs)


30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8327 - attrition_output_loss: 0.0000e+00 - dept_output_accuracy: 0.6521 - dept_output_loss: 0.5214 - loss: 0.5212 - val_attrition_output_accuracy: 0.7966 - val_attrition_output_loss: 0.0000e+00 - val_dept_output_accuracy: 0.6314 - val_dept_output_loss: 0.5296 - val_loss: 0.5382
Epoch 3/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8466 - attrition_output_loss: 0.0000e+00 - dept_output_accuracy: 0.6390 - dept_output_loss: 0.5359 - loss: 0.5358 - val_attrition_output_accuracy: 0.7966 - val_attrition_output_loss: 0.0000e+00 - val_dept_output_accuracy: 0.6314 - val_dept_output_loss: 0.5474 - val_loss: 0.5587
Epoch 4/100
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8436 - attrition_output_loss: 0.0000e+00 - dept_output_accuracy: 0.6545 - dept_output_loss: 0.5176 - loss: 0.5176 - val_attrition_output_accuracy: 0.7966 - val_attrition_output_loss: 0.0000e+00 - val_dept_out

In [151]:
# Evaluate the model with the testing data
test_loss, test_dept_loss, test_attrition_loss, test_dept_acc, test_attrition_acc = model.evaluate(
    X_test_scaled, 
    [y_test_dept, y_test_attrition],
    verbose=0
)


In [152]:
# Print the accuracy for both department and attrition
print("Model Evaluation on Test Data:")
print(f"Total test loss: {test_loss:.4f}")
print(f"Department test loss: {test_dept_loss:.4f}")
print(f"Attrition test loss: {test_attrition_loss:.4f}")
print(f"Department test accuracy: {test_dept_acc:.4f}")
print(f"Attrition test accuracy: {test_attrition_acc:.4f}")

Model Evaluation on Test Data:
Total test loss: 8.9982
Department test loss: 8.8402
Attrition test loss: 0.0000
Department test accuracy: 0.8673
Attrition test accuracy: 0.6667


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Accuracy might not be the best metric for this data, especially for the Attrition prediction. The reason behind this is there can be an imbalance of classes. For example if 85% of your employees stay then thre model predicts stay would have a 85% accuracy and would not be useful in prediction ppl that would leave.

2.For the output layers, I chose Softmax for department because it was a multi-class classification and it allows the model to assign probabilities to each department for a given input.
I chose Sigmoid for attrtion because it was a binary classification. The value of 0 and 1 works well when interpreting the probability of someone staying or leaving

3. Feature Engineering, Class Imbalance, Ensemble Methods